src/main/java/org/apache/mesos/scheduler/plan/DefaultBlock.java

```diff
@@ -0,0 +1,161 @@
+    }
+
+    private void setStatus(Status newStatus) {
+        Status oldStatus = status;
+        status = newStatus;
+        logger.info(getName() + ": changed status from: " + oldStatus + " to: " + newStatus);
+    }
+
+    @Override
+    public boolean isComplete() {
+        return isStatus(Status.COMPLETE);
+    }
+
+    @Override
+    public boolean isPending() {
+        return isStatus(Status.PENDING);
+    }
+
+    @Override
+    public boolean isInProgress() {
+        return isStatus(Status.IN_PROGRESS);
+    }
+
+    @Override
+    public Optional<OfferRequirement> start() {
+        return offerRequirementOptional;
+    }
+
+    @Override
+    public void updateOfferStatus(Optional<Collection<Protos.Offer.Operation>> operationsOptional) {
+        if (operationsOptional.isPresent()) {
+            setTaskIds(operationsOptional.get());
+            setStatus(Status.IN_PROGRESS);
+        }
+    }
+
+    @Override
+    public void restart() {
+        logger.warn("Restarting block: " + getName());
+        setStatus(Status.PENDING);
+    }
+
+    @Override
+    public void forceComplete() {
+        logger.warn("Forcing completion of block: " + getName());
+        setStatus(Status.COMPLETE);
+    }
+
+    @Override
+    public void update(Protos.TaskStatus status) {
+        if (!tasks.containsKey(status.getTaskId())) {
+            logger.warn(getName() + " ignoring irrelevant TaskStatus: " + status);
+            return;
+        }
+
+        if (isComplete()) {
+            logger.warn(getName() + " ignoring due to being Complete, TaskStatus: " + status);
+        }
+
+        switch (status.getState()) {
+            case TASK_RUNNING:
+                tasks.replace(status.getTaskId(), Status.COMPLETE);
+                break;
+            case TASK_ERROR:
+            case TASK_FAILED:
+            case TASK_FINISHED:
+            case TASK_KILLED:
+            case TASK_KILLING:
+                tasks.replace(status.getTaskId(), Status.ERROR);
+                // Retry the block because something failed.
+                setStatus(Status.PENDING);
+                break;
+            case TASK_STAGING:
+            case TASK_STARTING:
+                tasks.replace(status.getTaskId(), Status.IN_PROGRESS);
+                break;
+            default:
+                logger.warn("Failed to process unexpected state: " + status.getState());
+        }
+
+        for (Status taskStatus : tasks.values()) {
+            if (!taskStatus.equals(Status.COMPLETE)) {
+                // Keep and log current status
+                setStatus(this.status);
+                return;
+            }
+        }
+
+        setStatus(Status.COMPLETE);
+    }
+
+    @Override
+    public UUID getId() {
+        return id;
+    }
+
+    @Override
+    public String getMessage() {
+        return "Block: " + name;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Could use ToStringBuilder, like this: https://github.com/mesosphere/dcos-commons/blob/71de6c54646600ed510a0fcc94495b5df4960f5a/src/main/java/org/apache/mesos/offer/OfferRequirement.java#L115